src/main/common/filter.c

```diff
@@ -86,11 +86,8 @@ FAST_CODE float slewFilterApply(slewFilter_t *filter, float input)
 
 // get notch filter Q given center frequency (f0) and lower cutoff frequency (f1)
 // Q = f0 / (f2 - f1) ; f2 = f0^2 / f1
-float filterGetNotchQ(uint16_t centerFreq, uint16_t cutoff) {
-    const float f0sq = (float)centerFreq * centerFreq;
-    const float f1 = cutoff;
-
-    return (f1 * f0sq) / (f0sq - f1 * f1);
+float filterGetNotchQ(float centerFreq, float cutoffFreq) {
+    return centerFreq * cutoffFreq / (centerFreq - cutoffFreq) / (centerFreq + cutoffFreq);
```

Comment:
My fault .. I wrote wrong equation into Wolfram ...
http://www.wolframalpha.com/input/?i=f0%2F(f2+-+f1)+where+f2+%3D+f0%5E2+%2F+f1

It will be faster this way (avoiding division)
```
return centerFreq * cutoffFreq / ((centerFreq - cutoffFreq) * (centerFreq + cutoffFreq));
```
Another variant (multiplication and addition is one cycle, but MAC may be used here):
```
return centerFreq * cutoffFreq / (centerFreq * centerFreq - cutoffFreq * cutoffFreq);
```

Also may be better to return `fabs` of result, so both lower and higher cutoff can be specified

Reply:
Indeed, I didn't think about double division, thanks

Reply:
Question: Would gcc optimise this to have only one division on its own, or is that asking for too much?

Reply:
There are differences in generated assembler, but very small.

https://godbolt.org/g/mfTLB1

Reply:
@apoc : With optimization enabled: https://godbolt.org/g/cajgY8

Optimization is a bit limited - without -ffast-math, even division by constant can't be replaced with multiplication - result may be different due to rounding.

Maybe in this case multiplication can underflow or overflow where double division would work ...

Reply:
I had issues discovering this page :-D
@mikeller With -ffast-math it could happen, but compiler may still decide there's a better way to rewrite this, especially if it gets inlined. I wanted to compare generated assembly in BF, but forgot about it.
Will do that now.

Reply:
Here's how it looks for F3 in our code:
```
   0x08025f7c <+15944>:	vmov	s15, r3

				float notchQ = filterGetNotchQ(fftResult[axis].centerFreq, cutoffFreq);
   0x08025f80 <+15948>:	vmov	s14, r2

   0x08025f84 <+15952>:	vcvt.f32.s32	s15, s15	# s15	<- cutoffFreq
   0x08025f88 <+15956>:	vcvt.f32.u32	s0, s14		# s0	<- centerFreq

./src/main/common/filter.c:
				return centerFreq * cutoffFreq / (centerFreq * centerFreq - cutoffFreq * cutoffFreq);
   0x08025f8c <+15960>:	vnmul.f32	s1, s15, s15	# s1	<- -(cutoffFreq * cutoffFreq)
   0x08025f90 <+15964>:	vmul.f32	s15, s15, s0	# s15	<- cutoffFreq * centerFreq
   0x08025f94 <+15968>:	vfma.f32	s1, s0, s0		# s1	<- -(cutoffFreq * cutoffFreq) + centerFreq * centerFreq

   0x08025faa <+15990>:	vdiv.f32	s1, s15, s1		# s1	<- cutoffFreq * centerFreq / (-(cutoffFreq * cutoffFreq) + centerFreq * centerFreq)
```

Double-division option with `-ffast-math` indeed gets reduced to a single-division version.
Also ensured that `filterGetNotchQ` get's properly inlined for `gyroDataAnalyseUpdate`.
Inter-procedure analysis is so powerful these functions get spread across the final function as compiler sees fit, so it's pretty hard to collect the required disassembly manually :-)

Reply:
Same for F4 and F7.
@mikeller so the answer to your question is yes, with our compiler flags double division gets optimized out, but not as good as the fused-multiply-accumulate version.

Reply:
@DieHertz: My question was just out of interest, I did not mean to imply we should rely on the compiler to optimise - this isn't a case where the unoptimised form offers better readability.

Reply:
My study was out of interest as well, besides I wanted to ensure that FFT calculation inlines the function properly, as rav-rav used a macro instead of `filterNotchGetQ` before, probably because he wasn't sure it'll get inlined.
We should rely on the compiler in many cases IMO.

Reply:
With `-ffast-math` : https://godbolt.org/g/SrVxk2
gcc can infer equivalence of both versions ...

Reply:
@ledvinap yep, already tried it using your link before. It defines original in terms of `test2`.

Reply:
@apoc : ARM cortex is designed as quite predictable CPU - short pipeline, no prediction, in-order execution (except div/sqrt and lazy ISR register store).
Caching is available (+ flash 'accelerator'), but there is little compiler can do with it